unrpyc.py

```diff
@@ -395,49 +379,59 @@ def traverse(inpath):
+                baselist.append(elem)
 
     # Check if we actually have files. Don't worry about no parameters passed,
     # since ArgumentParser catches that
-    if not worklist:
+    if not baselist:
         print("Found no script files to decompile.")
         return
 
     # If a big file starts near the end, there could be a long time with only one thread running,
     # which is inefficient. Avoid this by starting big files first.
-    worklist.sort(key=lambda x: x.stat().st_size, reverse=True)
-    worklist = [(args, x) for x in worklist]
-
-    if args.processes > 1 and len(worklist) > 5:
-        with Pool(args.processes, sharelock, [printlock]) as pool:
-            results = pool.map(worker, worklist)
-    else:
-        results = list(map(worker, worklist))
-
-    if args.write_translation_file:
-        print(f'Writing translations to {args.write_translation_file}...')
-        translated_dialogue = {}
-        translated_strings = {}
-        good = 0
-        bad = 0
-        for result in results:
-            if not result:
-                bad += 1
-                continue
-            good += 1
-            translated_dialogue.update(pickle_loads(result[0]))
-            translated_strings.update(result[1])
-        with args.write_translation_file.open('wb') as out_file:
-            pickle_safe_dump((args.language, translated_dialogue, translated_strings), out_file)
-
-    else:
-        # Check per file if everything went well and report back
-        good = results.count(True)
-        bad = results.count(False)
+    baselist.sort(key=lambda x: x.stat().st_size, reverse=True)
+
+    while mp_runs != 0:
+        run_type = "basic decompiling"
+        if args.dump:
+            run_type = "astdump"
+        elif args.translate:
+            run_type = f'translation step {args.tl_step}'
+        print(f'> Multiprocessing run of type {run_type} starts.')
+
+        worklist = []
+        results = []
+        # if tl mode runs, worklist must be packed again to include new args
+        # baselist needs to stay unaltered so we use two names
+        worklist = [(args, x) for x in baselist]
+
+        if args.processes > 1 and len(worklist) > 5:
+            with Pool(args.processes, sharelock, [printlock]) as pool:
+                results = pool.map(worker, worklist)
+        else:
+            results = list(map(worker, worklist))
+
+        if args.tl_step == 1:
+            print('Sorting extracted translations.')
+            tl_dialogue = {}
+            tl_strings = {}
+            for entry in results:
+                if not entry:
+                    continue
+                tl_dialogue.update(pickle_loads(entry[0]))
+                tl_strings.update(entry[1])
+            args.translate2 = (tl_dialogue, tl_strings)
+            args.tl_step = 2
+
+        mp_runs -= 1
+
+    # Check per file if everything went well and report back
+    good = results.count(True)
+    bad = results.count(False)
```

Comment:
I'd prefer the translation run to be more compartmentalized. It's really not necessary combining them to use the same inner loop with a lot of conditionals inside them. Just run another loop first to collect the translation data, and then run the normal loop with some translation data.

Reply:
Well, that's another way. I look at it.